api.js

```diff
@@ -0,0 +1,92 @@
+"use strict";
+var debugEnabled = require("./config").debug;
+var log = require("color-logs")(debugEnabled, debugEnabled, __filename);
+
+var jwt = require("jsonwebtoken");
+
+class BaseApi {
+
+  constructor() {
+    this.req = null;
+    this.res = null;
+  }
+
+  success(res, data) {
+    if(!res) return;
+    return res.status(200).send({
+      success: true,
+      data: data
+    });
+  }
+
+  fail(res, code, message, data) {
+    return res.status(200).send({
+      success: false,
+      code: code,
+      message: message,
+      data: data
+    });
+  }
+
+  error(res, code, message) {
+    return res.status(401).send({
+      success: false,
+      message: message
+    });
+  }
+  exception(res, ex) {
+    return res.status(500).send({
+      success: false,
+      data: ex
+    });
+  }
+
+  permissionDenied(res) {
+    return res.status(401).send("permission denied");
+  }
+
+  isAuthenticated(req, res, next) {
+    let data = this.getToken(req, res);
+    log.info("checking authentication of", data);
+    if (data) {
+      next();
+    } else {
+      return this.permissionDenied(res);
+    }
+  }
+
+  getToken(req, res) {
+    let jwtSecret = require("./config").jwtSecret
+    let token = req.headers['x-access-token'] || req.headers['authorization'];
+    if(!token) {
+      this.permissionDenied(res);
+    }
+
+    if (token.startsWith('Bearer ')) {
+      // Remove Bearer from string
+      token = token.slice(7, token.length);
+    }
+
+    if (token) {
+      try {
+        var decoded = jwt.verify(token, jwtSecret);
+        return decoded;
+      } catch(e) {
+        log.error(e);
+        if (e.name == "TokenExpiredError") {
+          return this.fail(res, 4010, "Token Expired", {
+            message: e.message,
+            expiredAt: e.expiredAt
+          });
+        }
+        log.error(e);
+        return this.error(401, "Token is not valid");
```

Comment:
There is a bug in dealing with invalid token, it won't return 401 error.
For example when I run the app locally:
```
curl -XGET -H 'x-access-token: test' 'http://localhost:3003/me'  
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Error</title>
</head>
<body>
<pre>TypeError: res.status is not a function<br> &nbsp; &nbsp;at UsersController.error (/Users/testuser/Work/wcl-test/paulovelho/api.js:32:16)<br> &nbsp; &nbsp;at UsersController.getToken (/Users/testuser/Work/wcl-test/paulovelho/api.js:83:21)<br> &nbsp; &nbsp;at UsersController.isAuthenticated (/Users/testuser/Work/wcl-test/paulovelho/api.js:49:21)<br> &nbsp; &nbsp;at Layer.handle [as handle_request] (/Users/testuser/Work/wcl-test/paulovelho/node_modules/express/lib/router/layer.js:95:5)<br> &nbsp; &nbsp;at next (/Users/testuser/Work/wcl-test/paulovelho/node_modules/express/lib/router/route.js:137:13)<br> &nbsp; &nbsp;at Route.dispatch (/Users/testuser/Work/wcl-test/paulovelho/node_modules/express/lib/router/route.js:112:3)<br> &nbsp; &nbsp;at Layer.handle [as handle_request] (/Users/testuser/Work/wcl-test/paulovelho/node_modules/express/lib/router/layer.js:95:5)<br> &nbsp; &nbsp;at /Users/testuser/Work/wcl-test/paulovelho/node_modules/express/lib/router/index.js:281:22<br> &nbsp; &nbsp;at Function.process_params (/Users/testuser/Work/wcl-test/paulovelho/node_modules/express/lib/router/index.js:335:12)<br> &nbsp; &nbsp;at next (/Users/testuser/Work/wcl-test/paulovelho/node_modules/express/lib/router/index.js:275:10)<br> &nbsp; &nbsp;at module.exports (/Users/testuser/Work/wcl-test/paulovelho/services/connect.js:15:2)<br> &nbsp; &nbsp;at Layer.handle [as handle_request] (/Users/testuser/Work/wcl-test/paulovelho/node_modules/express/lib/router/layer.js:95:5)<br> &nbsp; &nbsp;at trim_prefix (/Users/testuser/Work/wcl-test/paulovelho/node_modules/express/lib/router/index.js:317:13)<br> &nbsp; &nbsp;at /Users/testuser/Work/wcl-test/paulovelho/node_modules/express/lib/router/index.js:284:7<br> &nbsp; &nbsp;at Function.process_params (/Users/testuser/Work/wcl-test/paulovelho/node_modules/express/lib/router/index.js:335:12)<br> &nbsp; &nbsp;at next (/Users/testuser/Work/wcl-test/paulovelho/node_modules/express/lib/router/index.js:275:10)</pre>
</body>
</html>
```

Reply:
I forgot to include `res` in this function call... =$

fixed.